quinn-proto/src/connection/streams/recv.rs

```diff
@@ -304,40 +285,43 @@ impl<'a> Chunks<'a> {
 
     fn finalize_inner(&mut self, drop: bool) -> ShouldTransmit {
         let state = mem::replace(&mut self.state, ChunksState::Finalized);
-        match state {
-            ChunksState::Readable(mut rs) => {
-                debug_assert!(!drop);
-                let should_transmit =
-                    Self::done(&mut rs, self.read, self.id, self.streams, self.pending);
-                self.streams.recv.insert(self.id, rs);
-                should_transmit
-            }
-            ChunksState::Finished => {
-                debug_assert!(!drop);
-                // MAX_DATA may need to be issued, but MAX_STREAM_DATA is pointless
-                let max_data = self.streams.add_read_credits(self.read);
-                self.pending.max_data |= max_data.0;
-                max_data
+        debug_assert!(
```

Comment:
This looks great! Do you think it's feasible/not too hard to do this change (centralize post-read logic) in like two or three steps? One step seems to be renaming the `Error` state, another to move logic from `next()` into `finalize_inner()`, another to inline `post_read()`, maybe more? As is, this is a bit hard to review, and given how finicky this has been/is turning out to be, I think it'd be good to be careful about changing it.

Reply:
Splitting went better than expected; should be much easier to digest now!

Reply:
Yup, much better!